### DISCUSSION TITLE
Kunena Discuss refactored

### DISCUSSION THREAD

**Discussion by Ruud68** [Kunena/Kunena-Forum#8942] [General]
Hi,
just did a complete overhaul of the discuss plugin here: https://github.com/Ruud68/Kunena-Addons/tree/refactored-j4

With these changes, Kunena Discuss is:

- a full blown J4 plugin
- utlilizes a J4 form for  posting a message (including form validations for the form fields): one advantage is that the form is now styled conform the site template!
- uses layouts ~~*~~
- bootstrap 5 tmpl files (tested on cassiopeia)

This is a working test / draft version.

~~* What I run into and is not fully working are the layouts: you can now select a layout in the plugin advanced parameters (currently: default or aurelia)~~
~~But because the plugin doesn't load the messages (topic items) via a Joomla / plugin layout the layouts only work partially. method KunenaForum::display('Topic', 'default', null, $params); will always load the default layout files (from the Kunena template) and NOT the configured layout files for the plugin. So even if you set aurelia in the plugin, it will load e.g. the aurelia_form.php, aurelia_login.php, etc. but not the aurelia.php / aurelia_left.php etc. files > it will always load the default.php and default_left.php files.~~

~~IMO it would be better if we could use the template files configured in the plugin and not use kunenaforum::display to display the messages / topics.~~

EDIT: refactored the template / layouts and removed the hardcoded Kunena template dependency for messages. Kunena Discuss can now have independent templates (tmpl) that are fully overridable via Joomla template overrides

Thoughts?

**Comment by Ruud68** [comment]
plugin is now fully functional, there where some issues with the captcha validation rule where the Joomla provided rule always validates when on an article page (which we where) and when articles parameters have captcha disabled (IMO this is a Joomla issue as it makes the captcha useless when used by e.g. a plugin or module on an article page).

**Comment by xillibit** [comment]
Hello,

The modules and  plugins uses the legacy part instead of using the HMVC way like it was done in Kunena. I will give a try

**Reply by Ruud68** [reply]
Top, the way I refactored doesn't uses the Kunena template at all: pure joomla :) If you need an installable package, please let me know. It installs as is: no cleanup of previous version so that may lead to issues > best to uninstall old one and install this one.

**Reply by xillibit** [reply]
I have tried your changes but now when add message with the form under the article it's not added and the topic related to the article isn't created

**Reply by rich20** [reply]
The Submit button does not send the message, it only refreshes the page. This is the problem.
If the setting "Create Topic on First Visit" the article is created in the forum. You can reply in the forum, but not at the article page.

**Reply by Ruud68** [reply]
> The Submit button does not send the message, it only refreshes the page. This is the problem. If the setting "Create Topic on First Visit" the article is created in the forum. You can reply in the forum, but not at the article page.

found the issue (was when posting with account), fixed now

**Comment by Ruud68** [comment]
Hi, working for me (using this in production now on multiple sites).
I noticed that the build process is broken, so I added a ./dist directory with an installable package (the one i am using now on production).
Just for testing.

**Reply by xillibit** [reply]
The build process works in the original repository, what is the issue with build process ?

**Reply by Ruud68** [reply]
> The build process works in the original repository, what is the issue with build process ?

BUILD FAILED
/home/ruud/GitHub/Kunena-Addons/plugins/content/kunenadiscuss/build/build.xml:99:61 Could not create task/type: 'filter'. Make sure that this class has been declared using taskdef / typedef.

Total time: 0.4080 seconds

Maybe due to other phing version? building kunenaforum works as expected

**Comment by Ruud68** [comment]
Ok, thanks for the explanation, never use that setting (only create on first reaction as otherwise my forum fills with 'empty' topics.
Will test and see why this is not working, not sure if the logic for this has changed

**Reply by Ruud68** [reply]
fixed :)

**Reply by rich20** [reply]
- [x] Success

**Reply by xillibit** [reply]
Can-you make a PR please against the K6.0 branch : https://github.com/Kunena/Kunena-Addons/tree/K6.0 ?

**Comment by Ruud68** [comment]
Done: https://github.com/Kunena/Kunena-Addons/pull/294
